src/behavior.ts

```diff
@@ -89,20 +93,56 @@ export abstract class Behavior<A> extends Reactive<A>
     return new ChainBehavior<A, B>(this, fn);
   }
   flatten: <B>(this: Behavior<Behavior<B>>) => Behavior<B>;
-  at(): A {
-    return this.state === State.Push ? this.last : this.pull();
+  at(t?: number): A {
+    if (this.state !== State.Push) {
+      const time = t === undefined ? tick() : t;
+      this.pull(time);
+    }
+    return this.last;
   }
-  push(a: any): void {
-    this.last = this.pull();
-    this.pushToChildren(this.last);
+  abstract update(t: number): A;
+  pushB(t: number): void {
+    if (this.state === State.Push) {
+      const newValue = this.update(t);
+      this.pulledAt = t;
+      if (this.last !== newValue) {
+        this.changedAt = t;
+        this.last = newValue;
+        this.pushToChildren(t);
+      }
+    }
   }
-  pull(): A {
-    return this.last;
+  pushToChildren(t: number): void {
+    for (const child of this.children) {
+      child.pushB(t);
+    }
+  }
+  pull(t: number): void {
+    let shouldRefresh = this.changedAt === undefined;
+    for (const parent of this.parents) {
+      if (isBehavior(parent)) {
+        if (parent.state !== State.Push && parent.pulledAt !== t) {
+          parent.pull(t);
+        }
+        shouldRefresh = shouldRefresh || parent.changedAt > this.changedAt;
+      }
+    }
+    if (shouldRefresh) {
+      this.refresh(t);
+    }
+    this.pulledAt = t;
+  }
+  refresh(t: number) {
+    const newValue = this.update(t);
+    if (newValue !== this.last) {
+      this.changedAt = t;
+      this.last = newValue;
+    }
   }
   activate(): void {
     super.activate();
     if (this.state === State.Push) {
-      this.last = this.pull();
+      this.refresh(tick());
```

Comment:
Activating a parent advances the global time? Should `activate` take the time as an argument instead?